### PR TITLE
Check TF ops for ONNX compliance

### PR DESCRIPTION
**Pull request by jplu**
# What does this PR do?

This PR aims to check if a model is compliant with ONNX Opset12 by adding a quick test and a script. The script is only for testing a saved model while the quick test aims to be run over a manually built graph. For now, only BERT is forced to be compliant with ONNX, but the test can be unlocked for any other model.

The logic can also be extended to any other framework/SDK we might think of, such as TFLite or NNAPI.

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
LGTM but let's wait for Morgan's review. Would also be helpful to know which models should have the test activated (GPT-2 for instance?)

**Comment by jplu** [comment]
That's the moment to create that list :)

**Comment by mfuntowicz** [comment]
Really like the idea of improving our ONNX compatibility in a more reliable way. 

In this sens, I'm not sure this is the easiest way for approaching the problem. 
May be it would be more suitable to just attempt to export the model through `keras2onnx` and report errors. This can also allow us to more easily test compatibility with various ONNX opset (10 is the minimum required).

Also, regarding the list of model to be supported I think we would like to have: 
- BERT
- GPT2
- BART (cc @Narsil wdyt?)

**Comment by jplu** [comment]
> In this sense, I'm not sure this is the easiest way for approaching the problem. 
May be it would be more suitable to just attempt to export the model through `keras2onnx` and report errors. This can also allow us to more easily test compatibility with various ONNX opset (10 is the minimum required).

I'm not in favour of running the export directly in the tests, as it is less flexible and not a compatible solution with other frameworks/SDKs. We can add any other opsets without problems with the proposed approach, but I don't think that going below the 12 is good idea. Only a few of the current models are compliant with opset < 12. Also, the default in the convert script is 11, not 10, so maybe we can propose opset 11 to be aligned. I think the proposed models below are compliant.

> Also, regarding the list of model to be supported I think we would like to have:
BERT
GPT2
BART (cc @Narsil wdyt?)

These three are ok on my side!

**Comment by jplu** [comment]
Add BART and GPT2 as a mandatory green test.

**Comment by mfuntowicz** [comment]
Testing ONNX operator support might be more complicated than this. 

Each operator in itself supports a set of different shape(s) as input/output combined with different data type and various dynamic axis support ... 

I would go for the easiest solution, well tested of using the official converter to report incompatibilities.

File: tests/test_modeling_tf_common.py
Comment (Narsil) [inline tests/test_modeling_tf_common.py:226]: Again not familiar with TF way of working, but the actual inputs in PT do change quite a bit the actual traced graph.

That means that `use_cache` graph vs `first_pass` graph look quite different. Also setting variable batch_length where it can be fixed (for instance input_ids is necessarily [B, 1] for `use_cache` graph on decoder) can link to greatly different performance later down the road.

What I'm trying to say is that this test will probably check that the Ops used in TF are valid for some ONNX opset, **it does not by any means that check it can/will export the best production ready graph.**

And the real hot path in production is almost always, decoder-only with `use_cache` (even input_ids [1, 1]) within a generation loop (I don't think TF has the generation loop optimized yet.)

**Review by Narsil — COMMENTED** [review]
I think this PR is fine, but I want to stress some points about the claims of those tests, which do not reflect (at least not to my knowledge) the complexity of optimizing a graph for inference.

It also feels odd to add many empty tests for many models that are not used.

As for BART, it's a good example of encoder-decoder so having it might help improve confidence in ONNX exportability but don't think it's required for this PR.

**Comment by jplu** [comment]
The problem with the solution to use the converter is that we cannot have the full list of incompatible operators, it will stop at the first encounter one, which would be too much annoying IMO. We should also assume that as long as the operator belongs to this list https://github.com/onnx/tensorflow-onnx/blob/master/support_status.md it is compliant. Until now, this assumption is true for all of our models.

Unless you know a case for which it is not true?

Also, I'm afraid to add a dependency to the onnxruntime would switch the quick test into a slow test, which reduces the traceability of a potential change that will break it.

File: tests/test_modeling_tf_common.py
Comment (jplu) [reply tests/test_modeling_tf_common.py:218]: This is easier to maintain than a dict. Also this list should be shared across the check script.

File: tests/test_modeling_tf_common.py
Comment (jplu) [reply tests/test_modeling_tf_common.py:224]: I don't understand what you mean. Do you have a specific example for which the will be an error?

File: tests/test_modeling_tf_common.py
Comment (jplu) [reply tests/test_modeling_tf_common.py:226]: Here we are not testing the graph, we are loading the entire list of operators, the graph here is not optimized. To give you an example, For BERT we load the > 5000 operators, while the optimised graph for inference is only around 1200 nodes. The role of this test is just to be sure to have the entire list of used operators inside the list proposed here https://github.com/onnx/tensorflow-onnx/blob/master/support_status.md

File: tests/test_modeling_tf_common.py
Comment (jplu) [reply tests/test_modeling_tf_common.py:224]: The list is not the same. The list in `onnx_compliancy` is bigger because the script checks from a SavedModel, and inside a SavedModel you have operators that are added that are specific to a SavedModel. Here we check from the graph created on the fly, not from a SavedModel, so the operators that are specific to a SavedModel are not needed here.

File: tests/test_modeling_tf_common.py
Comment (Narsil) [reply tests/test_modeling_tf_common.py:226]: Yes I know, I was just emphasizing it.
Also `optimized` graph of ONNX can only go so far. It cannot know about `past_values` if they are not passed within those dummy inputs. 

unoptimized small graph > optimized big graph

- big as in, large sequences length, not sheer node number
- Again talking about PT here, I didn't check with TF yet, but results are probably similar.

Comment (jplu) [reply tests/test_modeling_tf_common.py:226]: I think there is a misunderstanding, this test is only here to say "this TF op is also implemented in ONNX" nothing more. And not for testing if the optimized ONNX graph will work as expected or not.

Comment (Narsil) [reply tests/test_modeling_tf_common.py:226]: > I think there is a misunderstanding, this test is only here to say "this TF op is also implemented in ONNX" nothing more. 
There is no misunderstanding, I was trying to say  what you just said.

Comment (jplu) [reply tests/test_modeling_tf_common.py:226]: Ok, so if you are trying to say the same thing there is no problem then^^

**Comment by LysandreJik** [comment]
We can add a test depending on `keras2onnx` or `onnxruntime` with a `@require_onnx` decorator. If you decide to go down this road, according to the time spent doing those tests, we'll probably put them in the slow suite.

**Comment by jplu** [comment]
I  like the idea to add a decorator. I will add a slow test doing this in addition to the quick test.

**Comment by jplu** [comment]
I have reworked the quick test. Now, we can easily specify against which opset we want to test a model to be compliant. In the `onnx.json` file, all the operators are split in multiple opset, where each of them corresponds to the list of operators implemented in it. This should be way much easier to maintain and more flexible to use.

In addition to this I have added slow test that runs a complete pipeline of "keras model -> ONNX model -> optimized ONNX model -> quantized ONNX model".

**Comment by jplu** [comment]
As proposed by @mfuntowicz I switched the min required opset version from 12 to 10 for BERT, GPT2 and BART.

File: src/transformers/file_utils.py
Comment (LysandreJik) [inline src/transformers/file_utils.py:155]:
```suggestion
    logger.debug(f"Successfully imported onnx version {_onxx_version}")
```

**Review by LysandreJik — COMMENTED** [review]
Great, I like that you use `keras2onnx` directly.

Will wait for @mfuntowicz's review before merging.

- Do you have an idea of how long the slow tests take ?
- According to the information gathered, would it be possible (in a next PR) to have a doc referencing the opset compliancy/onnx support for each model?

**Comment by jplu** [comment]
> Do you have an idea of how long the slow tests take ?

It depends on the model size, BERT is around 1min while BART is a bit more than 10min.

> According to the information gathered, would it be possible (in a next PR) to have a doc referencing the opset compliancy/onnx support for each model?

Do you mean to have an entire page about ONNX? Or just to add a paragraph in the doc of every model about it?

**Comment by jplu** [comment]
@LysandreJik The issue with `TFGPT2ForSequenceClassification` is fixed, I just removed the occurrence of `tf.map_fn` to have something simpler. 👌

File: tests/test_modeling_tf_common.py
Comment (mfuntowicz) [inline tests/test_modeling_tf_common.py:267]: This will work only for model < 2Gb

File: tests/test_modeling_tf_common.py
Comment (mfuntowicz) [inline tests/test_modeling_tf_common.py:269]: You can remove this, no need to save this graph on disk.

File: tests/test_modeling_tf_common.py
Comment (mfuntowicz) [inline tests/test_modeling_tf_common.py:270]: Ok, it's the old way of quantizing the model, and not the one recommended anymore which might lead to strange result.

I would not try quantization here as it's hardware dependant and we have no control on the machine actually running this test.

**Review by mfuntowicz — APPROVED** [review]

File: tests/test_modeling_tf_common.py
Comment (jplu) [reply tests/test_modeling_tf_common.py:267]: The model configs in the test make the size of the models tiny. So as it is only for testing, this case should never happen.

**Comment by jplu** [comment]
LGTM on my side!

@LysandreJik I have fixed the issue with `TFGPT2ForSequenceClassification`, so now it is compliant with ONNX.

@mfuntowicz I should have addressed your comments, please double check ^^

**Comment by mfuntowicz** [comment]
LGTM 👍🏻

**Comment by jplu** [comment]
@LysandreJik Feel free to merge if the recent changes look ok for you!

**Review by LysandreJik — APPROVED** [review]
LGTM! The changes you applied for TF GPT-2 could also be applied to other causal models that have this approach in their sequence classification architecture (such as CTRL), but they're not ONNX compliant so this can be done another time.

**Comment by jplu** [comment]
@LysandreJik Yes, this is exactly that :) I plan to apply this update to the other causal models one by one 😉

**Pull request merged**
